### PR TITLE
add support for variable-length tuples and typing.Collection (#101)

### PR DESCRIPTION
**Pull request by neelmraman**
closes https://github.com/anntzer/defopt/issues/101

### CODE REVIEW

File: test_defopt.py [hunk 1]
```diff
@@ -1196,7 +1208,7 @@ def test_manual_version(self):
              self._assert_streams(stdout=r'\Afoo 42\n\Z'):
             defopt.run([], version='foo 42', argv=['--version'])
 
-    def test_no_version(self):
+    def test_moduleless(self):
```
Comment (neelmraman) [inline test_defopt.py:1211]: `test_no_version` was already used

File: lib/defopt.py [hunk 2]
```diff
@@ -91,6 +91,7 @@ def _ti_get_origin(tp):
         return {  # Py<3.7.
             typing.List: list,
             typing.Iterable: collections.abc.Iterable,
+            typing.Collection: collections.abc.Collection,
```
Comment (anntzer) [inline lib/defopt.py:94]: `getattr(typing, "Collection", object())`

File: lib/defopt.py [hunk 3]
```diff
@@ -568,8 +576,7 @@ def _update_help_string(action, *, show_defaults, show_types):
 
 def _is_list_like(type_):
     return (_ti_get_origin(type_)
-            in [list, collections.abc.Iterable, collections.abc.Sequence])
-
+            in [list, collections.abc.Iterable, collections.abc.Collection, collections.abc.Sequence])
```
Comment (anntzer) [inline lib/defopt.py:579]: need linewrap

File: lib/defopt.py [hunk 4]
```diff
@@ -991,11 +998,17 @@ def _make_literal_parser(literal, parsers, value=None):
             value, ', '.join(map(repr, map(str, _ti_get_args(literal))))))
 
 
-def _make_store_tuple_action_class(tuple_type, member_types, parsers):
-    parsers = [_get_parser(arg, parsers) for arg in member_types]
+def _make_store_tuple_action_class(tuple_type, member_types, parsers, *, is_variable_length=False):
+    if is_variable_length:
+        parsers = [_get_parser(member_types[0], parsers)]
```
Comment (anntzer) [inline lib/defopt.py:1003]: I think I would prefer
```python
if is_variable_length:
    parsers = itertools.repeat(_get_parser(member_types[0], parsers))
```
and don't modify parsers in `__call__` below.  (This way, at least in theory, one could reuse the same parser multiple times -- this is not currently possible in defopt, but one could easily imagine an API where an ArgumentParser instance is returned and the user can call `parse_args` on it multiple times.)

**Comment by anntzer** [comment]
Thanks, mostly looks good but a few things to fix.

File: lib/defopt.py (hunk 4 above)
Comment (neelmraman) [reply lib/defopt.py:1003]: Ah yeah good point

File: lib/defopt.py (hunk 2 above)
Comment (anntzer) [reply lib/defopt.py:94]: Because typing.Collection doesn't seem to exist on Py3.5?

Comment (neelmraman) [reply lib/defopt.py:94]: Ahhh yeah it looks like it was introduced in 3.6. Sorry about that!

File: lib/defopt.py [hunk 5]
```diff
@@ -91,6 +92,7 @@ def _ti_get_origin(tp):
         return {  # Py<3.7.
             typing.List: list,
             typing.Iterable: collections.abc.Iterable,
+            getattr(typing, "Collection", object()): collections.abc.Collection,
```
Comment (anntzer) [inline lib/defopt.py:95]: Looks like collection.abc.Collection also needs to be replaced by `getattr(collection.abc, "Collection", object())` again for Py3.5 support (below too).  (To be clear, on Py3.5, nothing will have an "origin" of `Collection`, so we can just replace that by an arbitrary object that is never referenced anywhere else.)

Comment (neelmraman) [reply lib/defopt.py:95]: Got it, and would you prefer `getattr(collection.abc, "Collection", object())` to something like
```python
try:
    from collections.abc import Collection
except ImportError:
    Collection = object()
```
and then just using `Collection` wherever `collections.abc.Collection` is referenced?

Comment (anntzer) [reply lib/defopt.py:95]: Actually that looks like a better solution, yes.

Comment (anntzer) [reply lib/defopt.py]: Tests need to be updated too (use `defopt.Collection` instead of `typing.Collection`).

Comment (neelmraman) [reply lib/defopt.py]: ah yeah whoops, will do

**Comment by anntzer** [comment]
I allowed myself to push a few minor fixes to your branch directly.  I also went back to `getattr(collections.abc, 'Collection', object())` because the construct actually only appears twice, and this makes it less likely to be confused with typing.Collection.

**Comment by anntzer** [comment]
Thanks for the PR!

**Pull request merged**